libsolutil/RationalVectors.h

```diff
@@ -0,0 +1,162 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolutil/LP.h>
+
+#include <libsolutil/Common.h>
+#include <libsolutil/CommonData.h>
+#include <libsolutil/StringUtils.h>
+#include <liblangutil/Exceptions.h>
+
+#include <range/v3/view/tail.hpp>
+#include <range/v3/algorithm/all_of.hpp>
+
+#include <optional>
+#include <stack>
+
+
+namespace solidity::util
+{
+
+using rational = boost::rational<bigint>;
+
+inline std::string toString(rational const& _value, size_t _paddedLength = 2)
+{
+	std::string result;
+	if (_value == rational(bigint(1) << 256))
+		result = "2**256";
+	else if (_value == rational(bigint(1) << 256) - 1)
+		result = "2**256-1";
+	else if (_value.denominator() == bigint(1))
+		result = _value.numerator().str();
+	else
+		result = to_string(_value);
+	if (result.length() < _paddedLength)
+		result = std::string(_paddedLength - result.length(), ' ') + result;
+	return result;
+}
+
+/// Creates a constraint vector "_factor * x__index"
+inline std::vector<rational> factorForVariable(size_t _index, rational _factor)
+{
+	std::vector<rational> result(_index + 1);
+	result[_index] = move(_factor);
+	return result;
+}
+
+inline rational const& get(std::vector<rational> const& _data, size_t _index)
+{
+	static rational const zero;
+	return _index < _data.size() ? _data[_index] : zero;
+}
+
+template <class T>
+void resizeAndSet(std::vector<T>& _data, size_t _index, T _value)
+{
+	if (_data.size() <= _index)
+		_data.resize(_index + 1);
+	_data[_index] = std::move(_value);
+}
+
+inline std::vector<rational>& operator/=(std::vector<rational>& _data, rational const& _divisor)
+{
+	for (rational& x: _data)
+		if (x.numerator())
```

Comment:
Why the condition?

Reply:
Optimization: A lot of entries in the vector will be zero, but the rational class does not optimize for it.